File: backend/services/workout_version_control/actions/rollback.py
"""
actions/rollback.py — Revert a workout plan to a previous version.

Restores the plan's exercise list from a stored snapshot and logs the
rollback as a new version history entry, maintaining a complete audit trail.
"""

from __future__ import annotations

from fastapi import HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from sqlalchemy.orm import selectinload

from backend.models.workout import WorkoutPlan, WorkoutExercise, WorkoutVersionHistory
from backend.services.workout_version_control.schemas import (
    RollbackRequest,
    WorkoutPlanRead,
)
from backend.services.workout_version_control import history as history_service
from backend.services.workout_version_control.actions.update import _snapshot_plan


async def rollback_workout_plan(
    db: AsyncSession,
    user_id: int,
    plan_id: int,
    payload: RollbackRequest,
) -> WorkoutPlanRead:
    """
    Revert a workout plan to the state captured in a previous version snapshot.

    Workflow:
        1. Verify ownership and that the plan is active.
        2. Load the snapshot for the requested ``target_version``.
        3. Snapshot the *current* state (for the new history entry).
        4. Delete current :class:`WorkoutExercise` rows and rebuild from snapshot.
        5. Increment ``version_number`` and restore ``name`` from snapshot.
        6. Record a rollback entry in :mod:`history` (type = ``"rollback"``).

    Args:
        db:       Async SQLAlchemy session.
        user_id:  ID of the authenticated user (ownership check).
        plan_id:  ID of the WorkoutPlan to roll back.
        payload:  Contains the ``target_version`` to restore.

    Returns:
        The restored :class:`WorkoutPlanRead` schema.

    Raises:
        HTTPException 404: Plan or target version not found / not owned.
        HTTPException 410: Plan has been soft-deleted.
        HTTPException 400: Attempting to roll back to the current version.
    """
    # --- 1. Fetch plan ---
    plan_result = await db.execute(
        select(WorkoutPlan)
        .options(selectinload(WorkoutPlan.exercises))
        .where(WorkoutPlan.id == plan_id, WorkoutPlan.user_id == user_id)
    )
    plan = plan_result.scalar_one_or_none()

    if not plan:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Workout plan {plan_id} not found or does not belong to you.",
        )
    if not plan.is_active:
        raise HTTPException(
            status_code=status.HTTP_410_GONE,
            detail=f"Workout plan {plan_id} is deleted and cannot be rolled back.",
        )
    if payload.target_version == plan.version_number:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Plan is already at version {payload.target_version}.",
        )

    # --- 2. Load target snapshot ---
    snap_result = await db.execute(
        select(WorkoutVersionHistory).where(
            WorkoutVersionHistory.workout_plan_id == plan_id,
            WorkoutVersionHistory.version_number == payload.target_version,
        )
    )
    history_entry = snap_result.scalar_one_or_none()

    if not history_entry:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=(
                f"Version {payload.target_version} not found for plan {plan_id}. "
                "Cannot roll back to a non-existent snapshot."
            ),
        )

    target_snapshot = history_entry.previous_snapshot

    # --- 3. Snapshot current state before overwriting ---
    current_snapshot = _snapshot_plan(plan)

    # --- 4. Replace exercise rows ---
    for ex in plan.exercises:
        await db.delete(ex)
    await db.flush()

    restored_exercises = [
        WorkoutExercise(
            workout_plan_id=plan.id,
            exercise_id=ex_data["exercise_id"],
            sets=ex_data.get("sets", 3),
            reps=ex_data.get("reps", 10),
            weight=ex_data.get("weight", 0.0),
            rest_seconds=ex_data.get("rest_seconds", 60),
            order_index=ex_data.get("order_index", 0),
        )
        for ex_data in target_snapshot.get("exercises", [])
    ]
    db.add_all(restored_exercises)

    # --- 5. Update plan metadata ---
    if "name" in target_snapshot:
        plan.name = target_snapshot["name"]
    plan.version_number += 1
    await db.flush()

    # --- 6. Record rollback in history ---
    await history_service.record_version(
        db=db,
        plan_id=plan.id,
        new_version=plan.version_number,
        change_summary=(
            f"Rolled back to version {payload.target_version} "
            f"(previous version was {current_snapshot['version_number']})."
        ),
        previous_snapshot=current_snapshot,
        change_type="rollback",
    )

    await db.commit()
    await db.refresh(plan)
    return WorkoutPlanRead.model_validate(plan)

File: backend/services/workout_version_control/actions/rollback.ts
/**
 * actions/rollback.ts — Revert a workout plan to a previous version snapshot.
 *
 * Restores the plan's exercise list from a stored WorkoutVersionHistory
 * snapshot, then records the rollback itself as a new version entry so the
 * full audit trail is never broken.
 */

import { DataSource } from "typeorm";
import { WorkoutPlan, WorkoutExercise, WorkoutVersionHistory } from "../../../models/workout";
import { RollbackRequest, WorkoutPlanRead } from "../schemas";
import { recordVersion } from "../history";
import { snapshotPlan } from "./update";
import { serializePlan } from "./create";

/**
 * Revert a workout plan to the state captured in a previous version snapshot.
 *
 * Workflow:
 *  1. Verify ownership and that the plan is active.
 *  2. Load the WorkoutVersionHistory entry for `targetVersion`.
 *  3. Snapshot the *current* state (stored as the new history entry's previous state).
 *  4. Delete current WorkoutExercise rows and rebuild from the target snapshot.
 *  5. Restore plan name from snapshot; increment `versionNumber`.
 *  6. Record a "rollback" version history entry.
 *
 * @param dataSource TypeORM DataSource (injected).
 * @param userId     ID of the authenticated user (ownership check).
 * @param planId     ID of the WorkoutPlan to roll back.
 * @param payload    Contains the `targetVersion` to restore.
 * @returns          The restored WorkoutPlanRead DTO with new version number.
 * @throws           404 if plan or target version not found; 410 if deleted; 400 if same version.
 */
export async function rollbackWorkoutPlan(
  dataSource: DataSource,
  userId: number,
  planId: number,
  payload: RollbackRequest
): Promise<WorkoutPlanRead> {
  const planRepo = dataSource.getRepository(WorkoutPlan);
  const exerciseRepo = dataSource.getRepository(WorkoutExercise);
  const histRepo = dataSource.getRepository(WorkoutVersionHistory);

  // ── 1. Fetch plan ──────────────────────────────────────────────────────────
  const plan = await planRepo.findOne({
    where: { id: planId, userId },
    relations: ["exercises"],
  });

  if (!plan) {
    const err: any = new Error(
      `Workout plan ${planId} not found or does not belong to you.`
    );
    err.statusCode = 404;
    throw err;
  }
  if (!plan.isActive) {
    const err: any = new Error(
      `Workout plan ${planId} is deleted and cannot be rolled back.`
    );
    err.statusCode = 410;
    throw err;
  }
  if (payload.targetVersion === plan.versionNumber) {
    const err: any = new Error(`Plan is already at version ${payload.targetVersion}.`);
    err.statusCode = 400;
    throw err;
  }

  // ── 2. Load target snapshot ────────────────────────────────────────────────
  const historyEntry = await histRepo.findOne({
    where: { workoutPlanId: planId, versionNumber: payload.targetVersion },
  });

  if (!historyEntry) {
    const err: any = new Error(
      `Version ${payload.targetVersion} not found for plan ${planId}. ` +
        "Cannot roll back to a non-existent snapshot."
    );
    err.statusCode = 404;
    throw err;
  }

  const targetSnapshot = historyEntry.previousSnapshot;

  // ── 3. Snapshot current state ──────────────────────────────────────────────
  const currentSnapshot = snapshotPlan(plan);

  // ── 4. Replace exercise rows ───────────────────────────────────────────────
  await exerciseRepo.delete({ workoutPlanId: plan.id });

  const restoredExercises = (
    (targetSnapshot.exercises ?? []) as Record<string, unknown>[]
  ).map((ex) =>
    exerciseRepo.create({
      workoutPlanId: plan.id,
      exerciseId: ex.exerciseId as number,
      sets: (ex.sets as number) ?? 3,
      reps: (ex.reps as number) ?? 10,
      weight: (ex.weight as number) ?? 0,
      restSeconds: (ex.restSeconds as number) ?? 60,
      orderIndex: (ex.orderIndex as number) ?? 0,
    })
  );
  await exerciseRepo.save(restoredExercises);

  // ── 5. Update plan metadata ────────────────────────────────────────────────
  if (targetSnapshot.name) plan.name = targetSnapshot.name as string;
  plan.versionNumber += 1;
  await planRepo.save(plan);

  // ── 6. Record rollback in history ─────────────────────────────────────────
  await recordVersion(
    dataSource,
    plan.id,
    plan.versionNumber,
    `Rolled back to version ${payload.targetVersion} ` +
      `(previous version was ${currentSnapshot.versionNumber}).`,
    currentSnapshot,
    "rollback"
  );

  // Re-fetch with full relations for the response
  const restored = await planRepo.findOne({
    where: { id: plan.id },
    relations: ["exercises", "exercises.exercise"],
  });

  return serializePlan(restored!);
}
